### ISSUE TITLE
Missing data sources for managed rulesets and rule ids for the new WAF

### ISSUE THREAD

**Issue by bartbriek** [cloudflare/terraform-provider-cloudflare#1943]
### Current Terraform and Cloudflare provider version

Terraform version: 1.3.1
Cloudflare provider: 3.24.0 (~> 3.0)

### Description

Add the moment there are no data sources available for the managed rulesets and rules for the new WAF. 
This makes the code less readable / maintainable. Requested change is to add data sources for this so you can loop over and filter them by name / category. This is available for the legacy waf_groups or waf_rules (https://registry.terraform.io/providers/cloudflare/cloudflare/latest/docs/data-sources/cloudflare_waf_groups) 

### Use cases

Making exceptions for specific rules for a specific url. 
Example:
URL A is getting false positives for some of the OWASP rules. Eg. sqli rules are triggered, but there is no sql database behind the service. So you could make exceptions specificly for the rules that are triggered. 

### Potential Terraform configuration

**Managed ruleset id**
```
data "cloudflare_ruleset" "owasp_ruleset" {
  filter {
    name   = ".*OWASP.*"
  }
}
```

**Managed rule id**
```
data "cloudflare_rule" "invalid_character" {
  filter {
    name = ".*920272.*"
  }
}
```

You can then more readable add exceptions to your managed ruleset. 
```
# Exceptions
  rules {
    action  = "skip"
    enabled = true

    action_parameters {
      rules = {
        "${data.owasp_ruleset.id}" = "${data.invalid_character.id},..."
      }
    }
```

Even better would be, to make the value for the rule id in the exception part of the `cloudflare_ruleset` resource a list instead of a comma separated list. You then get something like this:

```
"${data.owasp_ruleset.id}" = [
    "${data.invalid_character.id},",
    "${data.other_data_source.id}",
]
```

**Comment by fzipi360** [comment]
When I find myself curling for IDs... I remember we are losing the data resources we had previously.

**Comment by bartbriek** [comment]
Awesome! Will try this out in the coming weeks.

**Comment by bartbriek** [comment]
@jacobbednarz Thanks for this! How did you see this being implemented for getting the rules of a ruleset? Considering we need to skip certain rules in de managed rulesets? Right now I would need to do the following:

```

data "cloudflare_rulesets" "example" {
  zone_id = var.zone_id
  filter {
    name = ".*OWASP.*"
    
  }
  include_rules = true
}

specific_rule = [for rule in data.cloudflare_rulesets.example.rulesets[0].rules : rule.id if rule.id == "2380cd409b604c2a9273042f3eb29c4e"] # rule 942490
``` 

Or do I see something completely wrong?
TLDR: I can filter based on ruleset but not yet on rule within a ruleset base on a rule name correct?

**Comment by fzipi360** [comment]
This part is tricky. I could not solve it 100% yet. Each rule has versions also. So I wanted to get one rule. but ended up having two, version "79" and "81" of the same rule. Let me see if I have the code around...

**Comment by fzipi360** [comment]
Ok, found it.

Based on this code:
```terraform
locals {
  owasp_data_rules = flatten([
    {
      for r in data.cloudflare_rulesets.owasp.rulesets[0].rules :
      r.description => r...
    }
  ])
}

data "cloudflare_rulesets" "owasp" {
  zone_id       = var.zone_id
  include_rules = true

  filter {
    name = "Cloudflare OWASP Core Ruleset"
  }
}
```

I wanted to get the map keyed by description, where you have the rule ID that you can use. Because some of the rules are repeated, but with different version numbers, the map ends up having duplicated keys and you need the ellipsis so you create a map with a list for every repeated description.

This is what you get:
```
          + "942100: SQL Injection Attack Detected via libinjection"                                                                                        = [
              + {
                  + action                   = "score"
                  + action_parameters        = []
                  + description              = "942100: SQL Injection Attack Detected via libinjection"
                  + enabled                  = true
                  + exposed_credential_check = []
                  + expression               = ""
                  + id                       = "2e8e540411f841ffa5f2e05748b74690"
                  + last_updated             = "2023-01-24 10:04:44.629165 +0000 UTC"
                  + logging                  = []
                  + ratelimit                = []
                  + ref                      = "0bf8f28bdeda82ee424779e20f1567fd"
                  + version                  = "81"
                },
              + {
                  + action                   = "score"
                  + action_parameters        = []
                  + description              = "942100: SQL Injection Attack Detected via libinjection"
                  + enabled                  = true
                  + exposed_credential_check = []
                  + expression               = ""
                  + id                       = "f69971ea46aa4cc983f773bb06e082e4"
                  + last_updated             = "2023-01-24 10:04:44.629165 +0000 UTC"
                  + logging                  = []
                  + ratelimit                = []
                  + ref                      = "0bf8f28bdeda82ee424779e20f1567fd2"
                  + version                  = "79"
                },
            ]
```

So ideally you could be able to filter the rules by description, but you need/want to use the latest version there. Honestly, I don't really know if this is a bug or a feature.

**Comment by bartbriek** [comment]
Nice stuff, but i hope this becomes more easy in the future! Because I would rather stay away from writing for loops in our terraform configuration for every rule and just create a data source for each rule that we need.
